cli/commands/dump/dump.go

```diff
@@ -0,0 +1,165 @@
+		Action{Resource: "hook", Verb: VerbList, Namespaced: true},
+		Action{Resource: "mutator", Verb: VerbList, Namespaced: true},
+		Action{Resource: "role", Verb: VerbList, Namespaced: true},
+		Action{Resource: "role-binding", Verb: VerbList, Namespaced: true},
+		Action{Resource: "silenced", Verb: VerbList, Namespaced: true},
+	}
+)
+
+// Command dumps generic Sensu resources to a file or STDOUT.
+func Command(cli *cli.SensuCli) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "dump [RESOURCE TYPE],[RESOURCE TYPE]... [-f FILE]",
+		Short: "dump resources to a file or STDOUT",
+		RunE:  execute(cli),
+	}
+
+	helpers.AddAllNamespace(cmd.Flags())
+	_ = cmd.Flags().StringP("format", "", cli.Config.Format(), fmt.Sprintf(`format of data returned ("%s"|"%s")`, config.FormatWrappedJSON, config.FormatYAML))
+	_ = cmd.Flags().StringP("file", "f", "", "file to dump resources to")
+
+	return cmd
+}
+
+func execute(cli *cli.SensuCli) func(*cobra.Command, []string) error {
+	return func(cmd *cobra.Command, args []string) error {
+		if len(args) != 1 {
+			_ = cmd.Help()
+			return errors.New("invalid argument(s) received")
+		}
+
+		// get the configured format or the flag override
+		format := cli.Config.Format()
+		if flag := helpers.GetChangedStringValueFlag("format", cmd.Flags()); flag != "" {
+			format = flag
+		}
+		switch format {
+		case config.FormatYAML, config.FormatWrappedJSON:
+		default:
+			format = config.FormatYAML
+		}
+
+		// parse the comma separated resource types and match against the defined actions
+		var actions []Action
+		if args[0] == "all" {
+			actions = All
+		} else {
+			// check for duplicates first
+			types := strings.Split(args[0], ",")
+			for i := 0; i < len(types); i++ {
+				for v := 0; v < i; v++ {
+					if types[v] == types[i] {
+						return fmt.Errorf("duplicate resource type: %s", types[v])
+					}
+				}
+			}
+			// build actions for sensuctl
+			for _, t := range types {
+				length := len(actions)
+				for _, action := range All {
+					if t == action.Resource {
+						actions = append(actions, action)
+					}
+				}
+				if length == len(actions) {
+					return fmt.Errorf("invalid resource type: %s", t)
+				}
+			}
+		}
+
+		// iterate the matched actions and start building a sensuctl command
+		var out string
+		for _, a := range actions {
+			ctlArgs := []string{
+				a.Resource,
+				a.Verb,
+				"--format",
+				format,
+			}
+
+			// append --namespace or --all-namespaces flag if compatible with the resource type
+			if a.Namespaced {
+				if ok, err := cmd.Flags().GetBool(flags.AllNamespaces); err != nil {
+					return err
+				} else if ok {
+					ctlArgs = append(ctlArgs, "--all-namespaces")
+				} else {
+					ctlArgs = append(ctlArgs, "--namespace")
+					ctlArgs = append(ctlArgs, cli.Config.Namespace())
+				}
+			}
+
+			// execute the command and build wrapped-json or yaml lists
+			originalBytes, _ := exec.Command(os.Args[0], ctlArgs...).Output()
```

Comment:
This is a bit problematic because if there is an error executing the command (say, the connection to the server is interrupted), the dump command will continue on and silently discard information.

It would be nice to detect failure, and pass on the stderr output from the command if that happens.

I'd recommend something like 
```
originalBytes, err := exec.Command(os.Args[0], ctlArgs...).CombinedOutput()
if err != nil {
    // print output and exit
}
```

Reply:
I believe I originally implemented this with `CombinedOutput()` but ran into an issue with enterprise resources. A licensed user would have no problem using `sensuctl dump all` since they have access to the injected enterprise resources. An unlicensed user however, might get something like below, where the silently discarded information is invalid and irrelevant. Happy to bounce some ideas around about a solution to this use case.
```
---
Error: whatever error is thrown if an unlicensed user tries to get license info
---
type: CheckConfig
api_version: core/v2
metadata:
  name: default
  namespace: default
spec:
  command: echo default
...
```

Reply:
Yeah I see the problem.. how annoying. If commands returned different status codes based on their error, then we could differentiate, but they don't.

Fixing this problem would require changing the status codes we return on error, or reimplementing everything with API calls, so let's go with what you've done for now. :+1: